### PR TITLE
fix: execution of devenv version in read only directory

### LINKED ISSUE #1211: Cannot get version without writable directory
**Describe the bug**
one cannot run `devenv version` unless CWD is writable

**To reproduce**

```
cfc0bd017ecc:/tmp$ devenv version
devenv 1.0.5 (aarch64-linux)
cfc0bd017ecc:/tmp$ pwd
/tmp
cfc0bd017ecc:/tmp$ cd /
cfc0bd017ecc:/$ pwd
/
cfc0bd017ecc:/$ devenv version
thread 'main' panicked at devenv/src/main.rs:282:45:
Failed to create .devenv/gc directory: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Version**
 
`devenv 1.0.5 (aarch64-linux)`

### PR DESCRIPTION
**Pull request by k3yss**
resolves  #1211

@sandydoo

### CODE REVIEW

File: devenv/src/main.rs [hunk 1]
```diff
@@ -201,6 +201,9 @@ enum Commands {
     #[clap(hide = true)]
     Assemble,
 
+    #[clap(hide = true)]
+    CreateDirectories,
+
```
Comment (sandydoo) [inline devenv/src/main.rs:206]: We don't need a CLI subcommand for this, do we?
```suggestion
```

File: devenv/src/main.rs [hunk 2]
```diff
@@ -268,6 +271,7 @@ struct App {
     has_processes: Option<bool>,
     container_name: Option<String>,
     assembled: bool,
+    dirs_created: Option<bool>,
```
Comment (sandydoo) [inline devenv/src/main.rs:274]: A `bool` would do the job here. Is there a reason to to wrap it in an `Option`?
```suggestion
    dirs_created: bool,
```

File: devenv/src/main.rs [hunk 3]
```diff
@@ -407,6 +409,7 @@ fn main() -> Result<()> {
             InputsCommand::Add { name, url, follows } => app.inputs_add(&name, &url, &follows),
         },
         // hidden
+        Commands::CreateDirectories => app.create_directories(),
```
Comment (sandydoo) [inline devenv/src/main.rs:412]:
```suggestion
```

File: devenv/src/main.rs [hunk 4]
```diff
@@ -1079,6 +1082,35 @@ impl App {
         Ok(())
     }
 
+    fn create_directories(&mut self) -> Result<()> {
+        if self.dirs_created != Some(true) {
```
Comment (sandydoo) [inline devenv/src/main.rs:1086]:
```suggestion
        if !self.dirs_created {
```

File: devenv/src/main.rs [hunk 5]
```diff
@@ -1079,6 +1082,35 @@ impl App {
...
+            let devenv_state_hash = {
+                let mut hasher = sha2::Sha256::new();
+                hasher.update(self.devenv_dotfile.to_string_lossy().as_bytes());
+                let result = hasher.finalize();
+                hex::encode(result)
+            };
+            self.devenv_runtime =
+                Path::new(&self.devenv_tmp).join(format!("devenv-{}", &devenv_state_hash[..7]));
+
+            self.dirs_created = Some(true);
```
Comment (sandydoo) [inline devenv/src/main.rs:1109]:
```suggestion
            self.dirs_created = true;
```

File: devenv/src/main.rs [hunk 6]
```diff
@@ -288,50 +292,48 @@ fn main() -> Result<()> {
...
         cli,
         config,
         assembled: false,
         has_processes: None,
         logger,
         container_name: None,
-        devenv_root,
-        devenv_dotfile,
-        devenv_dot_gc,
+        dirs_created: None,
```
Comment (sandydoo) [inline devenv/src/main.rs:322]:
```suggestion
        dirs_created: false,
```

File: devenv/src/main.rs [hunk 7]
```diff
@@ -288,50 +292,48 @@ fn main() -> Result<()> {
...
-        let mut hasher = sha2::Sha256::new();
-        hasher.update(devenv_dotfile.to_string_lossy().as_bytes());
-        let result = hasher.finalize();
-        hex::encode(result)
-    };
-    let devenv_runtime = Path::new(&devenv_tmp).join(format!("devenv-{}", &devenv_state_hash[..7]));
+
+    // using dummy values for initialization
+    let devenv_root: PathBuf = "./".into();
+    let devenv_runtime = Path::new(&devenv_tmp).join(format!("devenv-{}", 1234.to_string()));
```
Comment (sandydoo) [inline devenv/src/main.rs:312]: Why do we need to fake these? Surely we can fetch the current directory in most situations?

As an aside, `PathBuf::from` is probably the more rusty way of initialising a new pathbuf.

**Review by sandydoo — CHANGES_REQUESTED** [review]

File: devenv/src/main.rs (hunk 2 above)
Comment (k3yss) [reply devenv/src/main.rs]: Yeah, you are right. I was just following similar patterns as `assemble`, fixed everything in the latest commit.

**Pull request merged**